### PR TITLE
netm: Add Netem as new action for applying delay, loss etc. at same time

### PR DESCRIPTION
**Pull request by at15**
UCP #345 

### What problem does this PR solve? 

Fix #345 

### What is changed and how does it work?

Add a new action called `netem` for networkchaos, when provided, the controller will merge all individual specs (delay, loss etc.) into one `Netem` RPC request.

### Check List 

Tests 

 - [ ] Unit test, seem we don't have any ... #263 
 - [ ] E2E test, it's WIP #322 
 - [x] Manual test (add detailed scripts or steps below) blocked by #352 

Code changes

 - Has Go code change

Side effects

 - [ ] Breaking backward compatibility (depends on how we want impl it ...)

Related changes

 - [ ] Need to update the documentation

### Does this PR introduce a user-facing change?:

```release-note
Add netem as new network chaos action. It supports applying delay, loss etc. at same time.
```

### CODE REVIEW

**Comment by at15** [comment]
I tested manually, deployed nginx, then applied the example (labels are changed). I saw ping would go high because of the injected delay. However, loss is added at same time but I didn't see it happening ...

I didn't touch the webhook, it seems a lot of validation logic need to be added there, now most time invalid spec result in warning from `kubectl get events` like `nextRecover shouldn't be later than nextStart`

```
64 bytes from 10.244.1.4: icmp_seq=55 ttl=64 time=0.060 ms
64 bytes from 10.244.1.4: icmp_seq=56 ttl=64 time=0.185 ms
64 bytes from 10.244.1.4: icmp_seq=57 ttl=64 time=0.045 ms
```

```yaml
apiVersion: pingcap.com/v1alpha1
kind: NetworkChaos
metadata:
  name: network-netem-example
  namespace: chaos-testing
spec:
  action: netem
  mode: one
  selector:
    labelSelectors:
      "app": "nginx"
  delay:
    latency: "90ms"
    correlation: "25"
    jitter: "90ms"
  loss:
    loss: "25"
    correlation: "25"
  duration: "10s"
  scheduler:
    cron: "@every 15s"
```

PTAL @cwen0

File: controllers/networkchaos/netem/types.go [hunk 1]
```diff
@@ -254,15 +266,47 @@ func (r *Reconciler) applyPod(ctx context.Context, pod *v1.Pod, networkchaos *v1
...
+		return nil, errors.New(invalidNetemSpecMsg)
+	}
+
+	merged := &pb.Netem{}
+	for _, spec := range emSpecs {
+		em, err := spec.ToNetem()
+		if err != nil {
+			return nil, err
+		}
+		utils.MergeNetem(merged, em)
```
Comment (YangKeao) [inline controllers/networkchaos/netem/types.go]:
```
merged = utils.MergeNetem(merged, em)
```

I think the source of your bug is here. @at15 

In the chaos-daemon log, it doesn't receive correct merged netem.

Comment (at15) [reply controllers/networkchaos/netem/types.go]: Thank!, I get it now I am not using my own function properly, it should be

`merged = utils.MergeNetem(merged, em)` ....

Comment (at15) [reply controllers/networkchaos/netem/types.go]: Should get fixed now

```
2020-03-23T06:17:10.002Z	INFO	chaos-daemon-server	Set netem	{"Request": "netem:<time:90000 jitter:90000 delay_corr:25 loss:25 loss_corr:25 > container_id:\"containerd://08311decd2944bf9ef43f2031acb475b66c7ef11ee4a726786884fad1ddd27b8\" "}

64 bytes from 10.244.1.4: icmp_seq=23 ttl=62 time=0.136 ms

46 packets transmitted, 40 received, 13% packet loss, time 45489ms
```

**Comment by YangKeao** [comment]
@cwen0 PTAL

File: controllers/twophase/types.go [hunk 2]
```diff
@@ -128,6 +128,12 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		}
 
 		nextRecover := now.Add(*duration)
+		// TODO(at15): this should be validated in webhook instead of throwing error at runtime.
```
Comment (cwen0) [inline controllers/twophase/types.go]: Can you file an issue for this?

Comment (at15) [reply controllers/twophase/types.go]: Filed https://github.com/pingcap/chaos-mesh/issues/373

**Comment by you06** [comment]
The points has been fixed. Congratulation! You have awarded a badge for usability challenge program! Please fill the form to get your reward! http://tidbcommunity.mikecrm.com/QMCv4QL

**Pull request merged**